kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/SchemaPrinter.kt

```diff
@@ -0,0 +1,290 @@
+import com.apurebase.kgraphql.schema.directive.Directive
+import com.apurebase.kgraphql.schema.introspection.TypeKind
+import com.apurebase.kgraphql.schema.introspection.__Described
+import com.apurebase.kgraphql.schema.introspection.__Directive
+import com.apurebase.kgraphql.schema.introspection.__InputValue
+import com.apurebase.kgraphql.schema.introspection.__Schema
+import com.apurebase.kgraphql.schema.introspection.__Type
+import com.apurebase.kgraphql.schema.introspection.typeReference
+import com.apurebase.kgraphql.schema.model.Depreciable
+
+data class SchemaPrinterConfig(
+    /**
+     * Whether to *always* include the schema itself. Otherwise, it will only be included
+     * if required by the spec.
+     */
+    val includeSchemaDefinition: Boolean = false,
+
+    /**
+     * Whether to include descriptions.
+     */
+    val includeDescriptions: Boolean = false,
+
+    /**
+     * Whether to include built-in directives.
+     */
+    val includeBuiltInDirectives: Boolean = false
+)
+
+class SchemaPrinter(private val config: SchemaPrinterConfig = SchemaPrinterConfig()) {
+    /**
+     * Returns the given [schema] in schema definition language (SDL). Types and fields are sorted
+     * ascending by their name and appear in order of their corresponding spec section, i.e.
+     *
+     *  - 3.3 Schema
+     *  - 3.4 Types
+     *    - Scalars
+     *    - Objects
+     *    - Interfaces
+     *    - Unions
+     *    - Enums
+     *    - Input Types
+     *  - 3.13 Directives
+     *
+     * If descriptions are included, they will use single quotes, broken up on newlines, e.g.
+     *
+     * ```
+     *     "This is a long comment,"
+     *     "spanning over multiple"
+     *     "lines."
+     *     scalar DescriptedScalar
+     * ```
+     */
+    fun print(schema: __Schema): String {
+        // All actual (non-introspection) types of the schema
+        val schemaTypes = schema.types.filterNot { it.isIntrospectionType() || it.name == null }.sortedByName()
+
+        // Schema
+        //   https://spec.graphql.org/draft/#sec-Root-Operation-Types.Default-Root-Operation-Type-Names
+        //   "The type system definition language can omit the schema definition when each root operation type uses its respective default root type name and no other type uses any default root type name."
+        //   "Likewise, when representing a GraphQL schema using the type system definition language, a schema definition should be omitted if each root operation type uses its respective default root type name and no other type uses any default root type name."
+        val schemaDefinition = if (includeSchemaDefinition(schema)) {
+            buildString {
+                appendLine("schema {")
+                val indentation = "  "
+                // The query root operation type must always be provided
+                appendDescription(schema.queryType, indentation)
+                appendLine("${indentation}query: ${schema.queryType.name}")
```

Comment:
Using `String::prependIndent` could be easier to read, so we are not mixing formatting with semantics

Reply:
I was looking for a better way but didn't find that function. Will refactor.